fairlearn/adversarial/_adversarial_mitigation.py

```diff
@@ -0,0 +1,841 @@
+        :code:`'auto'` indicates to infer the predictor function
+        from the distribution type of :code:`y`
+        The string-keyword indicates a
+        distribution type, see user guide for more information.
+        A callable should be a
+        function that maps the continuous output of the predictor model to
+        a discrete prediction.
+
+    predictor_optimizer : str, torch.optim, tensorflow.keras.optimizers, callable, default = 'Adam'
+        The optimizer class to use. If a string is passed instead, this must be
+        either "SGD" or "Adam". A corresponding SGD or Adam optimizer is
+        initialized with the model and given learning rate.
+        If an instance of a subclass of torch.optim.Optimizer
+        or tensorflow.keras.optimizers.Optimizer is passed, this
+        is used directly. If a callable :code:`fn` is passed, we call this
+        callable and pass our model, and set the result of this call
+        as the optimizer, so: :code:`predictor_optimizer=fn(predictor_model)`.
+
+
+    adversary_optimizer : str, torch.optim, tensorflow.keras.optimizers, callable, default = 'Adam'
+        The optimizer class to use. Similarly defined as
+        :code:`predictor_optimizer`
+
+    constraints : str, default = 'demographic_parity'
+        The fairness measure to optimize for. Must be either 'demographic_parity'
+        (Demographic Parity) or 'equalized_odds' (Equalized Odds).
+
+    y_transform : sklearn.base.TransformerMixin, default = fairlearn.adversarial.FloatTransformer("auto")
+        The preprocessor to use on the predictions :code:`y`.
+        The given transformer *must* map data
+        to a 2d ndarray containing only floats. Per default, we use a
+        FloatTransformer that maps strictly binary or categorical tables
+        to binary or
+        one-hot encodings, and it maps strictly continuous-valued (possible 2d)
+        to itself.
+
+    a_transform : sklearn.base.TransformerMixin, default = fairlearn.adversarial.FloatTransformer("auto")
+        The preprocessor to use on the :code:`sensitive_features`.
+        The given transformer *must* map data
+        to a 2d ndarray containing only floats. Per default, we use a
+        FloatTransformer that maps strictly binary or categorical tables
+        to binary or
+        one-hot encodings, and it maps strictly continuous-valued (possible 2d)
+        to itself.
+
+    learning_rate : float, default = 0.001
+        A small number greater than zero to set as initial learning rate
+
+    alpha : float, default = 1.0
+        A small number :math:`\alpha` as specified in the paper. It
+        is the factor that balances the training towards predicting :math:`Y`
+        (choose :math:`\alpha` closer to zero) or increasing fairness
+        (choose larger :math:`\alpha`).
+
+    epochs : int, default = 1
+        Number of epochs to train for.
+
+    batch_size : int, default = -1
```

Comment:
One element per batch? I think that is a bit uncommon, but I can see why that is better than everything in one batch.

Reply:
I was just thinking about what should be a sensible default behavior. If I understand correctly, with epoch=1 and batch_size=-1, there would be only one update, which doesn't seem like a good idea.

There's usually an interplay between batch_size and learning_rate, see, e.g., the plots in [Revisiting small batch training for deep neural networks](https://arxiv.org/pdf/1804.07612.pdf). My sense is that choosing a small value of batch_size is generally a good choice for optimization as long as the learning rate is also appropriately small. However, to leverage parallel execution, it can be beneficial to increase the batch_size and also appropriately adjust learning_rate.